### PR TITLE
Re-use work for is_non_loader_key

### PR DESCRIPTION
**Pull request by jeffwashington**
#### Problem
ledger-tool verify shows improvement of ~160ms during load due to reusing the work to determine is_non_loader_key
#### Summary of Changes
reuse previous work
Fixes #

### CODE REVIEW

**Comment by jeffwashington** [comment]
a different mainnet snapshot verify:
non_loader:
ledger processing timing ExecuteTimings load_us **12551601** execute_us 105063600 store_us 8229971 details ExecuteDetailsTimings serialize_us 29926055 create_vm_us 3774563 execute_us 20217460 deserialize_us 15490889 changed_account_count 79432 total_account_count 2076021 total_data_size 145048986363 data_size_changed 1702452282
ledger processed in 1 minute and 15 seconds. root slot is 73534483 2 forks at 73534515 73534516 with 33 frozen banks

master:
ledger processing timing ExecuteTimings load_us **12464855** execute_us 104144878 store_us 8081794 details ExecuteDetailsTimings serialize_us 29803771 create_vm_us 3708143 execute_us 20241295 deserialize_us 15416750 changed_account_count 79432 total_account_count 2076021 total_data_size 145048986363 data_size_changed 1702452282
ledger processed in 1 minute and 15 seconds. root slot is 73534483 2 forks at 73534515 73534516 with 33 frozen banks

**Comment by taozhu-chicago** [comment]
> a different mainnet snapshot verify:
> non_loader:
> ledger processing timing ExecuteTimings load_us **12551601** execute_us 105063600 store_us 8229971 details ExecuteDetailsTimings serialize_us 29926055 create_vm_us 3774563 execute_us 20217460 deserialize_us 15490889 changed_account_count 79432 total_account_count 2076021 total_data_size 145048986363 data_size_changed 1702452282
> ledger processed in 1 minute and 15 seconds. root slot is 73534483 2 forks at 73534515 73534516 with 33 frozen banks
> 
> master:
> ledger processing timing ExecuteTimings load_us **12464855** execute_us 104144878 store_us 8081794 details ExecuteDetailsTimings serialize_us 29803771 create_vm_us 3708143 execute_us 20241295 deserialize_us 15416750 changed_account_count 79432 total_account_count 2076021 total_data_size 145048986363 data_size_changed 1702452282
> ledger processed in 1 minute and 15 seconds. root slot is 73534483 2 forks at 73534515 73534516 with 33 frozen banks

Not familiar with above log, shouldn't `non_loader` be faster than `master`?

**Comment by jeffwashington** [comment]
> Not familiar with above log, shouldn't `non_loader` be faster than `master`?

Well, yes, it should ;-) That is hilarious. Back to the drawing board ;-)

**Comment by jeffwashington** [comment]
> > Not familiar with above log, shouldn't `non_loader` be faster than `master`?
> 
> Well, yes, it should ;-) That is hilarious. Back to the drawing board ;-) Or, really, take some more measurements and see what I missed. This is either irrelevant compared to noise or something...

There was variation in the measurements. I took [20 measurements](https://docs.google.com/spreadsheets/d/11Ro8Dsg_pqwpEBJRrn0MtkPmdyQSdHv9umZFKSRR2vU/edit?usp=sharing) with and without and averaged them.

**Comment by jeffwashington** [comment]
Also note this line (prior to these changes) showed up in ryo's perf measurement logs as well.

**Comment by taozhu-chicago** [comment]
Agree perf measure has margin, looking at the change, the performance gain should depends on the size of program_ids. But it should not degrade performance. So yeah, it looks good to me

File: sdk/program/src/message.rs [hunk 1]
```diff
@@ -240,6 +240,26 @@ impl Sanitize for Message {
         Ok(())
     }
 }
+pub struct MessageNonLoaderKeyHelper<'a> {
```
Comment (carllin) [inline sdk/program/src/message.rs:243]: nit: rename to `MessageProgramIdsCache`

**Review by carllin — APPROVED** [review]

**Pull request merged**